### PR TITLE
dcos-log v2 improvements 

### PR DESCRIPTION
**Pull request by mnaboka**
- address technical debt from previous review
- add 2 new endpoints for browsing / downloading  files in/from the sandbox
- fixing bugs found during CLI testing

### CODE REVIEW

File: dcos-log/api/v2/handlers.go [hunk 1]
```diff
@@ -38,75 +38,110 @@ const (
 	eventStreamContentType = "text/event-stream"
 )
 
+type errSetupFilesAPIReader struct {
+	msg  string
+	code int
+}
+
+// Error implements errors interface.
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: btw, you don't need to document exported methods for unexported types. the linter lets you get by without it.

File: dcos-log/api/v2/handlers.go [hunk 2]
```diff
@@ -38,75 +38,110 @@ const (
...
+			code: http.StatusInternalServerError,
+		}
 	}
 
 	client, ok := middleware.FromContextHTTPClient(req.Context())
 	if !ok {
-		logError(w, "invalid context, unable to retrieve an *http.Client object", http.StatusInternalServerError)
-		return
+		return nil, errSetupFilesAPIReader{
+			msg:  "invalid context, unable to retrieve an *http.Client object",
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: Here and elsewhere, you might be able to use the `%T` formatting directive here to avoid creating a string literal with the type you were expecting. So something like 

```
msg: fmt.Sprintf("invalid context, unable to retrieve %T object", client)
```

Since `client` in this case is associated with a type.

File: dcos-log/api/v2/handlers.go [hunk 3]
```diff
@@ -121,7 +156,7 @@ func filesAPIHandler(w http.ResponseWriter, req *http.Request) {
 		opts = append(opts, reader.OptLines(limit))
 	}
 
-	cursor := -1
+	var cursor = -1
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: just curious, why did you change this to a `var` declaration?

File: dcos-log/api/v2/handlers.go [hunk 4]
```diff
@@ -50,9 +52,9 @@ func (e errSetupFilesAPIReader) Error() string {
 
 // TODO: pass a request to logError to enhance logging. (url, query parameters, any relevant headers, and so on)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: We can remove this TODO yea?

File: dcos-log/api/v2/handlers.go [hunk 5]
```diff
@@ -280,13 +295,21 @@ func redirectURL(id *nodeutil.CanonicalTaskID, file, RawQuery string) (string, e
 
 	// take the last element
 	taskID := id.ContainerIDs[len(id.ContainerIDs)-1]
-	taskLogURL := fmt.Sprintf("%s/%s/logs/v2/task/frameworks/%s/executors/%s/runs/%s/", prefix, id.AgentID,
+	taskLogURL := fmt.Sprintf("%s/%s/logs/v2/task/frameworks/%s/executors/%s/runs/%s", prefix, id.AgentID,
 		id.FrameworkID, executorID, taskID)
 
 	if isPod {
-		taskLogURL += fmt.Sprintf("/tasks/%s/%s", id.ID, file)
+		taskLogURL += fmt.Sprintf("/tasks/%s", id.ID)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: might be worth using `path.Join` here and further down where we are building paths for the `taskLogURL`...

File: dcos-log/api/v2/handlers.go [hunk 6]
```diff
@@ -313,7 +348,7 @@ func discover(w http.ResponseWriter, req *http.Request) {
 
 	if taskID == "" {
 		logrus.Error("taskID is empty")
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: since `logError()` does logging to logrus, can we remove this line?

File: dcos-log/api/v2/handlers.go [hunk 7]
```diff
@@ -347,22 +382,22 @@ func discover(w http.ResponseWriter, req *http.Request) {
...
+		logError(w, req, errMsg, http.StatusInternalServerError)
 		return
 	}
 
 	http.Redirect(w, req, taskURL, http.StatusSeeOther)
 }
 
-func journalHandler(w http.ResponseWriter, r *http.Request) {
-	acceptHeader := r.Header.Get("Accept")
+func journalHandler(w http.ResponseWriter, req *http.Request) {
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: small nit: i feel like these changes where we're just renaming a parameter could be put into a separate commit to ease the burden on the reviewer. not a big deal tho.

File: dcos-log/api/v2/handlers.go [hunk 8]
```diff
@@ -510,11 +545,65 @@ func journalHandler(w http.ResponseWriter, r *http.Request) {
...
 			}
 		}
 	}
 
 }
+
+func browseFiles(w http.ResponseWriter, req *http.Request) {
+	r, err := setupFilesAPIReader(req, "/files/browse")
+	if err != nil {
+		e, ok := err.(errSetupFilesAPIReader)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: I see this pattern a lot, where you compare something to the `errSetupFilesAPIReader`. i wonder if we could break some of this out into a utility function to reduce the repetition?

File: dcos-log/api/v2/handlers.go [hunk 9]
```diff
@@ -510,11 +545,65 @@ func journalHandler(w http.ResponseWriter, r *http.Request) {
...
+	downloadResp, err := r.Download()
+	if err != nil {
+		logError(w, req, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer downloadResp.Body.Close()
+
+	for k, vs := range downloadResp.Header {
+		for _, v := range vs {
+			w.Header().Add(k, v)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: Just curious... why are we adding all of the request headers into the response headers?  do we not want to just copy some of them?

File: dcos-log/api/v2/handlers.go [hunk 10]
```diff
@@ -510,11 +545,65 @@ func journalHandler(w http.ResponseWriter, r *http.Request) {
...
+		}
+	}
+
+	io.Copy(w, downloadResp.Body)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: I'd check the err response here and log it

File: dcos-log/api/v2/routes.go [hunk 11]
```diff
@@ -11,21 +11,38 @@ import (
 
 // InitRoutes inits the v1 logging routes
 func InitRoutes(v2 *mux.Router, cfg *config.Config, client *http.Client, nodeInfo nodeutil.NodeInfo) {
-	taskLogHandler := http.HandlerFunc(filesAPIHandler)
-	wrappedTaskLogHandler := middleware.Wrapped(taskLogHandler, cfg, client, nodeInfo)
+	// browse sandbox files
+	wrappedBrowseFiles := middleware.Wrapped(http.HandlerFunc(browseFiles), cfg, client, nodeInfo)
+	v2.Path("/task/frameworks/{frameworkID}/executors/{executorID}/runs/{containerID}/files/browse").Handler(wrappedBrowseFiles).Methods("GET")
```
Comment (collinvandyck) [inline dcos-log/api/v2/routes.go]: Here and elsewhere, I see path vars expressed as string literals in the URI path, as well as in the handlers when they need to extract values from the URI path.  does it make sense to constantize some of these?  Maybe not, but the thought came to me so i thought i'd just express it here.

File: dcos-log/api/v2/handlers.go [hunk 12]
```diff
@@ -2,6 +2,7 @@ package v2
 
 import (
 	"context"
+	"errors"
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: Just as a general suggestion -- this is a pretty big commit that touches a number of things. I think in the future this could be broken up into smaller commits according to scope.

File: dcos-log/mesos/files/reader/read.go [hunk 13]
```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
...
-		// TODO: split this function into smaller ones to improve the readability.
-		for {
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
-			lines, delta, err := rm.read(ctx, offset, length, reverse)
-			if err != nil {
-				cancel()
-				return nil, err
-			}
+		err := calcOffset(offset, length, rm)
+		if err != nil && err != io.EOF {
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: why does `io.EOF` get a pass here? because we keep monitoring at some point and start returning data eventually?

File: dcos-log/mesos/files/reader/read.go [hunk 14]
```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
...
+			return nil, err
+		}
+	}
 
+	// guard against negative offset
+	if rm.offset < 0 {
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: should this be an error condition?

File: dcos-log/mesos/files/reader/read.go [hunk 15]
```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
...
-			length = chunkSize
-			offset -= chunkSize - delta
-			rm.offset = offset
-
-			// if the offset is 0 or negative value, the means we reached the top of the file.
-			// we can just set the offset to 0 and read the entire file
-			if offset < 0 {
-				rm.offset = 0
-				break
+		// if the required number of lines found, we need to calculate an offset
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: This kind of logic should probably be in a separate commit with an explanation

File: dcos-log/mesos/files/reader/read.go [hunk 16]
```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
...
+		if foundLines+len(lines) >= skip {
+			for i, skipped := 0, 0; skipped < skip; i++ {
+				if lines[i].Message != "" {
+					skipped++
+				}
+				rm.offset -= len(lines[i].Message) + 1
 			}
+			break
+		} else {
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: if you have a `break` above, do you need the `else` here?

File: dcos-log/mesos/files/reader/read.go [hunk 17]
```diff
@@ -349,9 +369,68 @@ start:
...
+	Size  uint64 `json:"size"`
+	UID   string `json:"uid"`
+}
+
+type mTime uint64
+
+// mtime field has a weird format with trailing .0
+// in order to unmarshal this value into json, we need to cut off the suffix .0 if it exists
+func (mt *mTime) UnmarshalJSON(data []byte) error {
+	sanitizedData := bytes.TrimSuffix(data, []byte(".0"))
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: do we know that the suffix will always be `.0` ? if not, it may make sense to trim everything after, and including the `.` rune.

File: dcos-log/mesos/files/reader/read.go [hunk 18]
```diff
@@ -349,9 +369,68 @@ start:
...
+	resp, err := rm.client.Get(newURL.String())
+	if err != nil {
+		return nil, fmt.Errorf("unable to make a GET request: %s. URL %s", err, newURL.String())
+	}
+	defer resp.Body.Close()
+
+	var files []SandboxFile
+
+	if err := json.NewDecoder(resp.Body).Decode(&files); err != nil {
+		return nil, fmt.Errorf("unable to decode files API response: %s. URL %s", err, newURL.String())
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: you can replace `newURL.String()` here with just `newURL` since you're using `%s`

File: dcos-log/mesos/files/reader/read.go [hunk 19]
```diff
@@ -349,9 +369,68 @@ start:
...
+func (rm ReadManager) Download() (*http.Response, error) {
+	v := url.Values{}
+	v.Add("path", filepath.Join(rm.sandboxPath, rm.file))
+
+	newURL := rm.readEndpoint
+	newURL.RawQuery = v.Encode()
+
+	req, err := http.NewRequest("GET", newURL.String(), nil)
+	if err != nil {
+		return nil, err
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: does it make sense to wrap the `err` here with context about `newURL` ?

File: dcos-log/mesos/files/reader/read_test.go [hunk 20]
```diff
@@ -21,9 +21,14 @@ five
 `)
 )
 
-func createHandler(data []byte, t *testing.T) http.HandlerFunc {
+func createHandler(data []byte, useResponse bool, t *testing.T) http.HandlerFunc {
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read_test.go]: i might rename `useResponse` here to `forceSendResponse` just to make the intent clear.

File: dcos-log/mesos/files/reader/read_test.go [hunk 21]
```diff
@@ -166,3 +172,96 @@ func TestEmptyData(t *testing.T) {
...
+"nlink":1,
+"path":"\/var\/lib\/mesos\/slave\/slaves\/f7df47cd-c82b-470d-b84b-4fc8611a3976-S1\/frameworks\/f7df47cd-c82b-470d-b84b-4fc8611a3976-0001\/executors\/instance-parent-pod.ed021ce3-dea8-11e7-9679-bef2db080897\/runs\/34ddc098-3310-4a9b-9f07-3d42e4c334bc\/tasks\/parent-pod.instance-ed021ce3-dea8-11e7-9679-bef2db080897.container-1\/one",
+"size":10,
+"uid":"root"
+}]`)
+
+	ts := httptest.NewServer(createHandler(sandboxResponse, false, t))
+	defer ts.Close()
+
+	client := &http.Client{}
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read_test.go]: small nit: you can use `http.DefaultClient` here instead.

File: dcos-log/api/v2/handlers.go (hunk 1 above)
Comment (mnaboka) [reply dcos-log/api/v2/handlers.go]: TIL

File: dcos-log/api/v2/handlers.go (hunk 2 above)
Comment (mnaboka) [reply dcos-log/api/v2/handlers.go]: TIL :)

File: dcos-log/api/v2/handlers.go (hunk 3 above)
Comment (mnaboka) [reply dcos-log/api/v2/handlers.go]: not sure, but good catch

File: dcos-log/api/v2/handlers.go (hunk 9 above)
Comment (mnaboka) [reply dcos-log/api/v2/handlers.go]: my idea was to give back the user the original response from mesos files API, which should contain the response headers. I did not find a good way to copy the entire header object, so i had to iterate over it

File: dcos-log/mesos/files/reader/read.go (hunk 13 above)
Comment (mnaboka) [reply dcos-log/mesos/files/reader/read.go]: when we read backwards, we will call rm.read() eventually which returns io.EOF, this is actually a bug fix for something that we found in CLI integration tests

File: dcos-log/mesos/files/reader/read.go (hunk 14 above)
Comment (mnaboka) [reply dcos-log/mesos/files/reader/read.go]: no, because we use a chunk with a fixed size.
Suppose we have a file that has only 100 bytes. A chunk is 65536, so when we read backwards the may endup with weird offset set to `-65436`, but that is actually OK, that just means we need to read the file from the beginning (offset = 0)

File: dcos-log/mesos/files/reader/read.go (hunk 16 above)
Comment (mnaboka) [reply dcos-log/mesos/files/reader/read.go]: yeah, i will replace break with return, it just means stop the loop

File: dcos-log/mesos/files/reader/read.go (hunk 17 above)
Comment (mnaboka) [reply dcos-log/mesos/files/reader/read.go]: yeah, it's kind of weird response, but it will always be `.0`

File: dcos-log/mesos/files/reader/read.go (hunk 14 above)
Comment (collinvandyck) [reply dcos-log/mesos/files/reader/read.go]: makes sense thanks

File: dcos-log/api/v2/handlers.go [hunk 22]
```diff
@@ -605,5 +602,8 @@ func downloadFile(w http.ResponseWriter, req *http.Request) {
 		}
 	}
 
-	io.Copy(w, downloadResp.Body)
+	_, err = io.Copy(w, downloadResp.Body)
+	if err != nil {
+		logError(w, req, err.Error(), http.StatusInternalServerError)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: if we couldn't write the resp to `w` does it make sense to logError here?  maybe it makes sense to just log the error without trying to write to `w`?

File: dcos-log/mesos/files/reader/read.go [hunk 23]
```diff
@@ -67,9 +74,9 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
 		return nil, err
 	}
 
-	sandboxPath := fmt.Sprintf("/var/lib/mesos/slave/slaves/%s/frameworks/%s/executors/%s/runs/%s/", agentID, frameworkID, executorID, containerID)
+	sandboxPath := path.Join("/var/lib/mesos/slave/slaves/", agentID, "/frameworks/", frameworkID, "/executors/", executorID, "/runs/", containerID)
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: Small nit, you should not need the leading and trailing `/` when doing a `path.Join` on any of the arguments.

File: dcos-log/mesos/files/reader/read.go [hunk 24]
```diff
@@ -67,9 +74,9 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
...
 	if taskPath != "" {
-		sandboxPath += fmt.Sprintf("tasks/%s/", taskPath)
+		sandboxPath = path.Join(sandboxPath, "tasks/"+taskPath+"/")
```
Comment (collinvandyck) [inline dcos-log/mesos/files/reader/read.go]: This could be `path.Join(sandboxPath, "tasks", taskPath)` if you don't need the trailing slash.

File: dcos-log/api/v2/handlers.go [hunk 25]
```diff
@@ -299,16 +297,16 @@ func redirectURL(id *nodeutil.CanonicalTaskID, file, RawQuery string, browse, do
 		id.FrameworkID, executorID, taskID)
 
 	if isPod {
-		taskLogURL += fmt.Sprintf("/tasks/%s", id.ID)
+		taskLogURL += path.Join("/tasks/", id.ID)
```
Comment (collinvandyck) [inline dcos-log/api/v2/handlers.go]: small nit, i don't think you need the trailing `/` on `"/tasks/"` ... same elsewhere when joining paths.

File: dcos-log/api/v2/routes.go [hunk 26]
```diff
@@ -9,40 +9,49 @@ import (
...
-	v2.Path("/task/frameworks/{frameworkID}/executors/{executorID}/runs/{containerID}/files/browse").Handler(wrappedBrowseFiles).Methods("GET")
-	v2.Path("/task/frameworks/{frameworkID}/executors/{executorID}/runs/{containerID}/tasks/{taskPath}/files/browse").Handler(wrappedBrowseFiles).Methods("GET")
+	v2.Path(taskBrowsePath).Handler(wrappedBrowseFiles).Methods("GET")
+	v2.Path(podBrowsePath).Handler(wrappedBrowseFiles).Methods("GET")
 
 	// task logs
 	wrappedTaskLogHandler := middleware.Wrapped(http.HandlerFunc(filesAPIHandler), cfg, client, nodeInfo)
-	v2.Path("/task/frameworks/{frameworkID}/executors/{executorID}/runs/{containerID}/{file}").Handler(wrappedTaskLogHandler).Methods("GET")
-	v2.Path("/task/frameworks/{frameworkID}/executors/{executorID}/runs/{containerID}/tasks/{taskPath}/{file}").Handler(wrappedTaskLogHandler).Methods("GET")
+	v2.Path(taskPath + "/{file}").Handler(wrappedTaskLogHandler).Methods("GET")
```
Comment (collinvandyck) [inline dcos-log/api/v2/routes.go]: Here and elsewhere in `InitRoutes` we could use `path.Join` to be consistent, but it's no big deal.
